### PR TITLE
Concept Coach Task

### PR DESCRIPTION
**Pull request by Dantemss**
- Removed CC assistant
- Added CCTask migration/model
- Added CCTask get/create routines and domain routine
- Real get CC task endpoint
- Existing specs
- Made CC spaced practice ignore assignments that have not been started
- New specs

### CODE REVIEW

File: app/controllers/api/v1/cc/tasks_controller.rb [hunk 1]
```diff
@@ -12,64 +12,45 @@ class Api::V1::Cc::TasksController < Api::V1::ApiController
...
-  def create_fake_concept_coach_task
-    task =  Tasks::BuildTask[
-              task_type: :concept_coach,
-              title: 'Dummy task title',
-              description: 'Dummy task description',
-              opens_at: 1000.days.ago,
-              feedback_at: 1000.days.ago]
+    # If page_models.size > 1, the user is in 2 courses with the same CC book (not allowed)
+    page_model = page_models.order(:created_at).last
+    return head(:unprocessable_entity) if page_model.blank?
```
Comment (jpslav) [inline app/controllers/api/v1/cc/tasks_controller.rb]: Can you document this error and what it means in the apipie docs? (no CC for this user and this book)

File: app/controllers/api/v1/cc/tasks_controller.rb [hunk 2]
```diff
@@ -12,64 +12,45 @@ class Api::V1::Cc::TasksController < Api::V1::ApiController
...
 
-  def create_fake_concept_coach_task
-    task =  Tasks::BuildTask[
-              task_type: :concept_coach,
-              title: 'Dummy task title',
-              description: 'Dummy task description',
-              opens_at: 1000.days.ago,
-              feedback_at: 1000.days.ago]
+    # If page_models.size > 1, the user is in 2 courses with the same CC book (not allowed)
+    page_model = page_models.order(:created_at).last
```
Comment (jpslav) [inline app/controllers/api/v1/cc/tasks_controller.rb]: If `page_models` guaranteed to have 0 or 1 entries, why the `.order(:created_at).last`?

**Comment by jpslav** [comment]
We should maybe discuss this with @kevinburleigh75, but I'm fairly sure we don't want to pick the spaced practice problems when the CC task is created.  The reason being that the CC widget will request and retrieve the CC task whenever a student is logged in and they open a module in the book.  So if a student just clicks all over the book to check it out), the spaced practice history may be all messed up and the spaced practice exercises selected will definitely not reflect what they've actually read.

I believe the better course of action would be to populate the task with placeholders and then populate the placeholders after the first CC question is answered (likely in a background job).

File: app/routines/get_concept_coach.rb [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+    translations: { outputs: { type: :verbatim } },
+    as: :create_cc_task
+
+  uses_routine Tasks::CreateTasking,
+    translations: { outputs: { type: :verbatim } },
+    as: :create_tasking
+
+  uses_routine AddSpyInfo, as: :add_spy_info
+
+  uses_routine GetCourseEcosystem, as: :get_course_ecosystem
```
Comment (jpslav) [inline app/routines/get_concept_coach.rb]: Unused

File: app/routines/get_concept_coach.rb [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+  uses_routine GetHistory, as: :get_history
+
+  protected
+
+  def exec(role:, page:)
```
Comment (jpslav) [inline app/routines/get_concept_coach.rb]: The assumption in this signature is that someone else has done the work to verify that this role is allowed to have a CC for this page.  Would it make sense to move the logic from the controller that figures this out into this routine, and make the routine have this signature:

```ruby
def exec(user:, cnx_book_id:, cnx_page_id:)
```

This way there's no assumption about what was done before.  Also this gets a lot of code out of the controller.  

If the code does move here...

* I might recommend breaking it up into more private methods. 
* You'll have to replace the controller's `return head(:unprocessable_entity) if page_model.blank?` with a fatal_error, and then check for that error in the controller.

File: app/controllers/api/v1/cc/tasks_controller.rb (hunk 1 above)
Comment (jpslav) [inline app/controllers/api/v1/cc/tasks_controller.rb]: See you've already done this in the other PR

File: app/controllers/api/v1/cc/tasks_controller.rb (hunk 2 above)
Comment (Dantemss) [inline app/controllers/api/v1/cc/tasks_controller.rb]: It's only guaranteed by the fact that the enrollment process doesn't allow it (or won't allow it). There's no guarantee on the database etc.

File: app/routines/get_concept_coach.rb (hunk 4 above)
Comment (Dantemss) [inline app/routines/get_concept_coach.rb]: OK guess I'll combine the PR's because this will make error handling a nightmare...

**Comment by Dantemss** [comment]
Going to use https://github.com/openstax/tutor-server/pull/771. Closing this one.

**Pull request closed without merging**